src/language/dafnyInstallation.ts

```diff
@@ -1,477 +1,163 @@
-  const suffix = getDafnyPlatformSuffix();
-  return `${baseUri}/${tag}/dafny-${version}-x64-${suffix}.zip`;
-}
-
-export class DafnyInstaller {
-  public constructor(
-    private readonly context: ExtensionContext,
-    private readonly statusOutput: OutputChannel
-  ) {}
-
-  public isLatestKnownLanguageServerOrNewer(version: string): boolean {
-    if(version === LanguageServerConstants.UnknownVersion) {
-      this.writeStatus('failed to resolve the installed Dafny version');
-      return true;
-    }
-    const givenParts = version.split('.');
-    const latestVersion = LanguageServerConstants.LatestVersion;
-    const latestParts = latestVersion.split('.');
-    for(let i = 0; i < Math.min(givenParts.length, latestParts.length); i++) {
-      const given = givenParts[i];
-      const latest = latestParts[i];
-      if(given < latest) {
-        this.writeStatus(`the installed Dafny version is older than the latest: ${version} < ${latestVersion}`);
-        return false;
-      }
-      if(given > latest) {
-        this.writeStatus(`the installed Dafny version is newer than the latest: ${version} > ${latestVersion}`);
-        return true;
-      }
+  private async ensureDafnyIsInstalled(): Promise<string> {
+    const { path: dotnetExecutable } = await getDotnetExecutablePath();
+    const toolVersion = await this.getDotnetToolVersion();
+    const localToolPath = path.join(this.context.extensionPath, `out/resources/${toolVersion}/`);
+    if(fs.existsSync(localToolPath)) {
+      return localToolPath;
     }
-    this.writeStatus(`the installed Dafny version is the latest known: ${version} = ${latestVersion}`);
-    return true;
-  }
 
-  public async install(): Promise<boolean> {
-    this.statusOutput.show();
-    this.writeStatus('Starting Dafny installation');
     try {
-      await this.cleanInstallDir();
-      if(os.type() === 'Darwin' && os.arch() !== 'x64') {
-        // Need to build from source and move all files from Binary/ to the out/resource folder
-        this.writeStatus(`Found a non-supported architecture OSX:${os.arch()}. Going to install from source.`);
-        return await this.installFromSource();
-      } else {
-        const archive = await this.downloadArchive(await getDafnyDownloadAddress(this.context), 'Dafny');
-        await this.extractArchive(archive, 'Dafny');
-        await workspace.fs.delete(archive, { useTrash: false });
-        this.writeStatus('Dafny installation completed');
-        return true;
-      }
+      await mkdir(localToolPath, { recursive: true });
+      await execFileAsync(dotnetExecutable, [ 'new', 'tool-manifest' ], { cwd: localToolPath });
+      this.statusOutput.show();
+      window.showInformationMessage(Messages.Installation.Start);
+      this.writeStatus(Messages.Installation.Start);
+      await execFileAsync(dotnetExecutable, [ 'tool', 'install', 'Dafny', '--version', toolVersion ], { cwd: localToolPath });
```

Comment:
Wow it blows my mind. You just get the dotnet tool install instead of fetching binaries. Very clever.

Reply:
Well.. except it doesn't work well since Z3 is not provided. Leaving this code unused for now.

Reply:
Oh no! Another great reason to want to depend on Z3 directly.